{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# A Study of Transcription and Its Affects\n",
    "---\n",
    "**Contributors:** John M. Letey (John.Letey@colorado.edu), David A. Knox (David.Knox@colorado.edu)\n",
    "\n",
    "**Notes:** To get to my GitHub repository, go to this [website](https://github.com/JohnLetey/A-Study-of-Transcription-and-Its-Affects). For instructions on how to use my GitHub repository, go [here](https://github.com/JohnLetey/A-Study-of-Transcription-and-Its-Affects/blob/current/instructions.md).\n",
    "\n",
    "**Dependencies:** My Jupyter Notebook uses the [biopython](https://github.com/biopython/biopython) package, which you can easily download using the built in interface in the Anaconda Navigator. Look at my instructions [here](https://github.com/JohnLetey/A-Study-of-Transcription-and-Its-Affects/blob/current/instructions.md) on how to download biopython. I also have the dependency of having the package [pandas](https://github.com/pandas-dev/pandas), which should come pre-installed when you download Anaconda.\n",
    "\n",
    "This project follows the below pipeline.\n",
    "\n",
    "![pipeline](../Pictures/pipeline.png?raw=true)\n",
    "\n",
    "I have broken this project into to two major portions. The first portion calculates the hits (weak and strong transcription factor binding sights) for a specified chromosome (or all of the chromosomes) and transcription factor. The second portion analyses these hits. Before we talk about these two parts more in depth, I'm going to give you a brief introduction on why I'm doing this.\n",
    "\n",
    "## Introduction\n",
    "---\n",
    "The fundamental component of life, whether we're talking about animals or humans or anything, is the cell. The cell is the most complex and unique part of any living being. As humans, some of us wonder what goes on inside such a fundamental component of us. That when I started wondering, is there a pattern? In all cells, the DNA goes through a process called transcription which converts it into RNA (which then the cell uses to make proteins). While doing this, there are transcription factors, which the whole point of there existence is to bind to the DNA. Each transcription factor has a unique target string which it is looking for in the DNA sequence. \n",
    "\n",
    "## Imports\n",
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import Bio\n",
    "from Bio import SeqIO\n",
    "import pandas as pd\n",
    "from time import time\n",
    "import matplotlib.pyplot as plt\n",
    "%matplotlib inline\n",
    "import numpy as np\n",
    "import os"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Part 1: Finding the Hits\n",
    "---\n",
    "To find the hits, we have to have the three components\n",
    "\n",
    "1. Get the chromosomes from a specific fasta file (using the Python package biopython to read the file into a pandas dataframe).\n",
    "2. Get the PSSMs for all possible transcription factors from a specific tamo file\n",
    "3. Implement a function that calculates the output of a PSSM given a specific sequence\n",
    "\n",
    "When I am done implementing this, I'll move on to finding the hits.\n",
    "\n",
    "<!--<span style='color:blue'>some blue text</span>-->\n",
    "\n",
    "Let's implement a function that gets all of the data from a specified fasta file and puts it into a pandas dataframe using the Python module biopython."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def getCHRs(filename, numUseless):\n",
    "    # Get the data from the fasta file\n",
    "    with open(filename) as fasta_file:\n",
    "        identifiers = []\n",
    "        seq = []\n",
    "        for seq_record in SeqIO.parse(fasta_file, 'fasta'):\n",
    "            identifiers.append(seq_record.id)\n",
    "            seq.append(str(seq_record.seq))\n",
    "    # Put the data into a dataframe\n",
    "    s1 = pd.Series(identifiers[:len(identifiers)-numUseless], name='chromosome')\n",
    "    s2 = pd.Series(seq[:len(seq)-numUseless], name='sequence')\n",
    "    fastaDF = pd.DataFrame(dict(chromosome=s1, sequence=s2))\n",
    "    # Return the dataframe\n",
    "    return fastaDF"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now that I've defined the function that reads in the entire fasta file and puts it into a dataframe, let's call that function on my specific data, and also show what the dataframe looks like. I also need to take away some of the data that I'm not going to use, specifically, chrmt and 2-micron."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>CCACACCACACCCACACACCCACACACCACACCACACACCACACCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr2</td>\n",
       "      <td>AAATAGCCCTCATGTACGTCTCCTCCAAGCCCTGTTGTCTCTTACC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr3</td>\n",
       "      <td>CCCACACACCACACCCACACCACACCCACACACCACACACACCACA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr4</td>\n",
       "      <td>ACACCACACCCACACCACACCCACACACACCACACCCACACACCAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr5</td>\n",
       "      <td>CGTCTCCTCCAAGCCCTGTTGTCTCTTACCCGGATGTTCAACCAAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr6</td>\n",
       "      <td>GATCTCGCAAGTGCATTCCTAGACTTAATTCATATCTGCTCCTCAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr7</td>\n",
       "      <td>CCACACCCACACACACCACACCCACACCCACACACTACCCTAACAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr8</td>\n",
       "      <td>CCCACACACACCACACCCACACACCACACCCACACTTTTCACATCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr9</td>\n",
       "      <td>CACACACACCACACCCACACCACACCACACCACACCCACACCCACA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr10</td>\n",
       "      <td>CCCACACACACACCACACCCACACCCACACACACCACACCCACACA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chr11</td>\n",
       "      <td>CACCACACCCACACACCACACCCACACACACACCACACCCACACAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chr12</td>\n",
       "      <td>CACACACACACACCACCCACACACCACACCACACACCACACCACAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chr13</td>\n",
       "      <td>CCACACACACACCACACCCACACCACACCCACACCCACACCCACAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>chr14</td>\n",
       "      <td>CCGGCTTTCTGACCGAAATTAAAAAAAAAAAATGAAAATGAAACCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>chr15</td>\n",
       "      <td>ACACCACACCCACACCACACCCACACCCACACACCACACCCACACA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>chr16</td>\n",
       "      <td>AAATAGCCCTCATGTACGTCTCCTCCAAGCCCTGTTGTCTCTTACC...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chromosome                                           sequence\n",
       "0        chr1  CCACACCACACCCACACACCCACACACCACACCACACACCACACCA...\n",
       "1        chr2  AAATAGCCCTCATGTACGTCTCCTCCAAGCCCTGTTGTCTCTTACC...\n",
       "2        chr3  CCCACACACCACACCCACACCACACCCACACACCACACACACCACA...\n",
       "3        chr4  ACACCACACCCACACCACACCCACACACACCACACCCACACACCAC...\n",
       "4        chr5  CGTCTCCTCCAAGCCCTGTTGTCTCTTACCCGGATGTTCAACCAAA...\n",
       "5        chr6  GATCTCGCAAGTGCATTCCTAGACTTAATTCATATCTGCTCCTCAA...\n",
       "6        chr7  CCACACCCACACACACCACACCCACACCCACACACTACCCTAACAC...\n",
       "7        chr8  CCCACACACACCACACCCACACACCACACCCACACTTTTCACATCT...\n",
       "8        chr9  CACACACACCACACCCACACCACACCACACCACACCCACACCCACA...\n",
       "9       chr10  CCCACACACACACCACACCCACACCCACACACACCACACCCACACA...\n",
       "10      chr11  CACCACACCCACACACCACACCCACACACACACCACACCCACACAC...\n",
       "11      chr12  CACACACACACACCACCCACACACCACACCACACACCACACCACAC...\n",
       "12      chr13  CCACACACACACCACACCCACACCACACCCACACCCACACCCACAC...\n",
       "13      chr14  CCGGCTTTCTGACCGAAATTAAAAAAAAAAAATGAAAATGAAACCC...\n",
       "14      chr15  ACACCACACCCACACCACACCCACACCCACACACCACACCCACACA...\n",
       "15      chr16  AAATAGCCCTCATGTACGTCTCCTCCAAGCCCTGTTGTCTCTTACC..."
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get the data from the fasta file that I'm working with and take away two of those\n",
    "fastaDF = getCHRs('Data/SGDv3.fasta', 2)\n",
    "# Get the amount of chromosomes\n",
    "countCHR = fastaDF.count()\n",
    "numOfCHR = countCHR['chromosome']\n",
    "# Show the dataframe\n",
    "fastaDF"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I've read in all of the data I'm going to use from a specific fasta file and put it into a dataframe. Now I need to read in the all of the PSSMs from a specified tamo file into a dataframe."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def getPSSMs(filename):\n",
    "    # Get the PSSMs from the tamo file\n",
    "    tamo_file = open(filename)\n",
    "    tamo_data = []\n",
    "    line = tamo_file.readline()\n",
    "    while line:\n",
    "        tamo_data.append(line[:len(line)-1])\n",
    "        line = tamo_file.readline()\n",
    "    TF = []\n",
    "    PSSM = []\n",
    "    revPSSM = []\n",
    "    for i in range(len(tamo_data)):\n",
    "        if tamo_data[i][:6] == 'Source':\n",
    "            pssm = pd.DataFrame(pd.Series(['A', 'C', 'T', 'G'], name='Letter'))\n",
    "            REVpssm = pd.DataFrame(pd.Series(['A', 'C', 'T', 'G'], name='Letter'))\n",
    "            revpssm = []\n",
    "            lenOfPSSM = int((len(tamo_data[i-16]) - 3)/10) - 1\n",
    "            firstLine = tamo_data[i-16]\n",
    "            secondLine = tamo_data[i-15]\n",
    "            thirdLine = tamo_data[i-14]\n",
    "            fourthLine = tamo_data[i-13]\n",
    "            for j in range(lenOfPSSM):\n",
    "                value1 = float(firstLine[(j*10 + 6):(j*10 + 12)])\n",
    "                value2 = float(secondLine[(j*10 + 6):(j*10 + 12)])\n",
    "                value3 = float(thirdLine[(j*10 + 6):(j*10 + 12)])\n",
    "                value4 = float(fourthLine[(j*10 + 6):(j*10 + 12)])\n",
    "                revpssm.append([value1, value2, value3, value4])\n",
    "                pssm.insert(j+1, str(j), pd.Series([value1, value2, value3, value4]))\n",
    "            temp = revpssm\n",
    "            revpssm = []\n",
    "            for j in range(4):\n",
    "                List = []\n",
    "                for k in range(len(temp)):\n",
    "                    List.append(temp[k][j])\n",
    "                revpssm.append(List)\n",
    "            for j in range(4):\n",
    "                revpssm[j] = revpssm[j][::-1]\n",
    "            temp = revpssm[0]\n",
    "            revpssm[0] = revpssm[2]\n",
    "            revpssm[2] = temp\n",
    "            temp = revpssm[1]\n",
    "            revpssm[1] = revpssm[3]\n",
    "            revpssm[3] = temp\n",
    "            temp = revpssm\n",
    "            revpssm = []\n",
    "            for j in range(len(temp[0])):\n",
    "                List = []\n",
    "                for k in range(4):\n",
    "                    List.append(temp[k][j])\n",
    "                revpssm.append(List)\n",
    "            for j in range(lenOfPSSM):\n",
    "                REVpssm.insert(j+1, str(j), pd.Series(revpssm[j]))\n",
    "            TF.append(tamo_data[i][9:])\n",
    "            PSSM.append(pssm)\n",
    "            revPSSM.append(REVpssm)\n",
    "    # Put the data into a dataframe\n",
    "    s1 = pd.Series(TF, name='TF')\n",
    "    s2 = pd.Series(PSSM, name='ForwardsPSSM')\n",
    "    s3 = pd.Series(revPSSM, name='ReversePSSM')\n",
    "    pssms = pd.DataFrame(dict(TF=s1, ForwardsPSSM=s2, ReversePSSM=s3))\n",
    "    # Return the dataframe\n",
    "    return pssms"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's get the dataframe that contains all of the PSSMs for each transcription factor from my specific tamo file, and also show what the dataframe looks like. To access any one PSSM and it's reverse compliment, look up the corresponding index of the transcription factor you want (let's call that number `value`), them plug that value into the Python statement\n",
    "```python\n",
    "tamoDF['ForwardsPSSM'][value] # Get the forwards PSSM\n",
    "tamoDF['ReversePSSM'][value]  # Get the reverse compliment PSSM\n",
    "```\n",
    "If you execute the above line, it should give you your PSSM for the corresponding transcription factor you wanted."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ForwardsPSSM</th>\n",
       "      <th>ReversePSSM</th>\n",
       "      <th>TF</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Letter      0       1      2      3      4  ...</td>\n",
       "      <td>Letter      0       1      2      3      4  ...</td>\n",
       "      <td>HSF1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Letter      0      1      2      3      4   ...</td>\n",
       "      <td>Letter      0      1      2       3      4  ...</td>\n",
       "      <td>RPN4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Letter      0       1       2       3       ...</td>\n",
       "      <td>Letter      0       1       2       3       ...</td>\n",
       "      <td>THI2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Letter      0      1      2    3    4    5  ...</td>\n",
       "      <td>Letter      0      1    2    3    4    5    ...</td>\n",
       "      <td>UGA3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Letter      0      1      2      3      4   ...</td>\n",
       "      <td>Letter      0      1      2      3      4   ...</td>\n",
       "      <td>SMP1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Letter      0       1       2       3       ...</td>\n",
       "      <td>Letter       0      1       2       3       ...</td>\n",
       "      <td>RTG3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Letter      0      1      2       3      4  ...</td>\n",
       "      <td>Letter      0      1      2      3       4  ...</td>\n",
       "      <td>CIN5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Letter      0      1      2      3      4   ...</td>\n",
       "      <td>Letter      0      1      2      3      4   ...</td>\n",
       "      <td>DAL82</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Letter      0      1      2      3\n",
       "0      A ...</td>\n",
       "      <td>Letter      0      1      2      3\n",
       "0      A ...</td>\n",
       "      <td>DAL80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Letter      0      1      2      3      4   ...</td>\n",
       "      <td>Letter      0      1      2      3      4   ...</td>\n",
       "      <td>DAL81</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Letter      0      1      2      3      4   ...</td>\n",
       "      <td>Letter      0      1      2      3      4   ...</td>\n",
       "      <td>REB1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Letter       0       1       2       3      ...</td>\n",
       "      <td>Letter      0      1      2      3      4   ...</td>\n",
       "      <td>PUT3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Letter      0      1       2       3       4...</td>\n",
       "      <td>Letter      0      1      2       3       4 ...</td>\n",
       "      <td>BAS1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Letter      0      1      2      3      4   ...</td>\n",
       "      <td>Letter      0      1      2      3      4   ...</td>\n",
       "      <td>YAP3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Letter      0      1      2      3      4   ...</td>\n",
       "      <td>Letter      0      1      2      3      4   ...</td>\n",
       "      <td>SIP4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Letter      0      1      2      3      4   ...</td>\n",
       "      <td>Letter      0      1      2      3      4   ...</td>\n",
       "      <td>SOK2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Letter      0      1      2      3      4   ...</td>\n",
       "      <td>Letter      0      1      2      3      4   ...</td>\n",
       "      <td>NDD1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Letter      0      1      2      3      4   ...</td>\n",
       "      <td>Letter      0       1      2       3      4 ...</td>\n",
       "      <td>FKH1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Letter      0       1       2       3      4...</td>\n",
       "      <td>Letter      0       1      2       3       4...</td>\n",
       "      <td>SKO1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Letter      0      1      2      3      4   ...</td>\n",
       "      <td>Letter      0      1       2       3      4 ...</td>\n",
       "      <td>FKH2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Letter      0      1      2       3      4  ...</td>\n",
       "      <td>Letter      0      1      2       3      4  ...</td>\n",
       "      <td>CBF1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Letter      0       1       2       3       ...</td>\n",
       "      <td>Letter       0       1       2       3      ...</td>\n",
       "      <td>GCR1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Letter      0       1       2       3       ...</td>\n",
       "      <td>Letter       0       1       2       3      ...</td>\n",
       "      <td>GCR2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Letter       0       1       2       3      ...</td>\n",
       "      <td>Letter       0       1       2       3      ...</td>\n",
       "      <td>YOX1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Letter      0      1      2      3      4   ...</td>\n",
       "      <td>Letter      0      1      2      3      4   ...</td>\n",
       "      <td>HAP1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>Letter      0       1       2       3       ...</td>\n",
       "      <td>Letter       0       1       2       3      ...</td>\n",
       "      <td>OPI1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>Letter      0      1      2      3      4   ...</td>\n",
       "      <td>Letter      0      1      2      3      4   ...</td>\n",
       "      <td>CRZ1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>Letter      0      1      2      3      4   ...</td>\n",
       "      <td>Letter      0      1      2      3      4   ...</td>\n",
       "      <td>ABF1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>Letter      0      1      2      3      4\n",
       "0 ...</td>\n",
       "      <td>Letter      0      1      2      3      4\n",
       "0 ...</td>\n",
       "      <td>YDR026C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>Letter      0       1      2      3      4  ...</td>\n",
       "      <td>Letter      0      1      2      3       4  ...</td>\n",
       "      <td>SUT1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>94</th>\n",
       "      <td>Letter      0      1      2      3      4   ...</td>\n",
       "      <td>Letter      0      1      2      3      4   ...</td>\n",
       "      <td>RPH1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>Letter      0      1      2      3      4   ...</td>\n",
       "      <td>Letter      0      1      2      3      4   ...</td>\n",
       "      <td>ASH1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>Letter       0       1       2      3       ...</td>\n",
       "      <td>Letter       0       1       2      3       ...</td>\n",
       "      <td>XBP1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>Letter      0      1      2      3      4   ...</td>\n",
       "      <td>Letter      0      1      2      3      4   ...</td>\n",
       "      <td>IME1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>Letter      0      1       2      3       4 ...</td>\n",
       "      <td>Letter      0      1      2       3       4 ...</td>\n",
       "      <td>SNT2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>Letter      0      1      2      3      4   ...</td>\n",
       "      <td>Letter      0      1      2      3      4   ...</td>\n",
       "      <td>MAC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>Letter      0      1      2      3      4   ...</td>\n",
       "      <td>Letter      0      1      2      3      4   ...</td>\n",
       "      <td>MATA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>Letter      0      1      2      3      4\n",
       "0 ...</td>\n",
       "      <td>Letter      0      1      2      3      4\n",
       "0 ...</td>\n",
       "      <td>GZF3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td>Letter      0      1      2      3      4   ...</td>\n",
       "      <td>Letter      0      1      2      3      4   ...</td>\n",
       "      <td>NRG1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103</th>\n",
       "      <td>Letter      0      1      2      3      4   ...</td>\n",
       "      <td>Letter      0      1      2      3      4   ...</td>\n",
       "      <td>YML081W</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104</th>\n",
       "      <td>Letter      0      1      2      3      4\n",
       "0 ...</td>\n",
       "      <td>Letter      0      1      2      3      4\n",
       "0 ...</td>\n",
       "      <td>YHP1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105</th>\n",
       "      <td>Letter    0      1      2      3      4\n",
       "0   ...</td>\n",
       "      <td>Letter      0      1      2      3    4\n",
       "0   ...</td>\n",
       "      <td>ADR1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106</th>\n",
       "      <td>Letter      0      1      2      3      4\n",
       "0 ...</td>\n",
       "      <td>Letter      0      1      2      3      4\n",
       "0 ...</td>\n",
       "      <td>GTS1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>Letter      0      1      2       3      4  ...</td>\n",
       "      <td>Letter      0      1      2       3      4  ...</td>\n",
       "      <td>UME6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>108</th>\n",
       "      <td>Letter      0      1      2    3    4    5  ...</td>\n",
       "      <td>Letter      0      1    2    3    4    5    ...</td>\n",
       "      <td>GAL80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>Letter      0      1       2       3      4 ...</td>\n",
       "      <td>Letter      0       1      2       3      4 ...</td>\n",
       "      <td>AZF1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110</th>\n",
       "      <td>Letter       0       1       2       3      ...</td>\n",
       "      <td>Letter      0       1       2       3       ...</td>\n",
       "      <td>LEU3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111</th>\n",
       "      <td>Letter      0      1      2      3       4  ...</td>\n",
       "      <td>Letter      0      1       2       3      4 ...</td>\n",
       "      <td>MCM1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>112</th>\n",
       "      <td>Letter      0       1       2      3       4...</td>\n",
       "      <td>Letter       0       1       2       3      ...</td>\n",
       "      <td>MET32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>113</th>\n",
       "      <td>Letter      0      1       2       3       4...</td>\n",
       "      <td>Letter       0       1       2       3      ...</td>\n",
       "      <td>MET31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114</th>\n",
       "      <td>Letter       0       1       2       3      ...</td>\n",
       "      <td>Letter       0       1       2       3      ...</td>\n",
       "      <td>SNF1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>115</th>\n",
       "      <td>Letter      0      1      2      3\n",
       "0      A ...</td>\n",
       "      <td>Letter      0      1      2      3\n",
       "0      A ...</td>\n",
       "      <td>HAP5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>116</th>\n",
       "      <td>Letter      0       1       2       3       ...</td>\n",
       "      <td>Letter      0      1      2      3      4   ...</td>\n",
       "      <td>HAP4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117</th>\n",
       "      <td>Letter      0      1      2      3      4   ...</td>\n",
       "      <td>Letter      0      1      2      3      4   ...</td>\n",
       "      <td>IXR1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118</th>\n",
       "      <td>Letter      0      1      2      3      4   ...</td>\n",
       "      <td>Letter      0      1      2      3      4   ...</td>\n",
       "      <td>SFL1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>119</th>\n",
       "      <td>Letter      0      1      2      3\n",
       "0      A ...</td>\n",
       "      <td>Letter      0      1      2      3\n",
       "0      A ...</td>\n",
       "      <td>HAP3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120</th>\n",
       "      <td>Letter      0      1      2      3      4   ...</td>\n",
       "      <td>Letter      0      1      2      3      4   ...</td>\n",
       "      <td>HAP2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121</th>\n",
       "      <td>Letter      0      1      2      3       4  ...</td>\n",
       "      <td>Letter       0       1      2      3       4...</td>\n",
       "      <td>RAP1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>122</th>\n",
       "      <td>Letter      0      1      2      3      4\n",
       "0 ...</td>\n",
       "      <td>Letter      0      1      2      3      4\n",
       "0 ...</td>\n",
       "      <td>MOT3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>123</th>\n",
       "      <td>Letter      0      1      2      3      4   ...</td>\n",
       "      <td>Letter      0      1      2      3      4   ...</td>\n",
       "      <td>SKN7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>124 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          ForwardsPSSM  \\\n",
       "0      Letter      0       1      2      3      4  ...   \n",
       "1      Letter      0      1      2      3      4   ...   \n",
       "2      Letter      0       1       2       3       ...   \n",
       "3      Letter      0      1      2    3    4    5  ...   \n",
       "4      Letter      0      1      2      3      4   ...   \n",
       "5      Letter      0       1       2       3       ...   \n",
       "6      Letter      0      1      2       3      4  ...   \n",
       "7      Letter      0      1      2      3      4   ...   \n",
       "8      Letter      0      1      2      3\n",
       "0      A ...   \n",
       "9      Letter      0      1      2      3      4   ...   \n",
       "10     Letter      0      1      2      3      4   ...   \n",
       "11     Letter       0       1       2       3      ...   \n",
       "12     Letter      0      1       2       3       4...   \n",
       "13     Letter      0      1      2      3      4   ...   \n",
       "14     Letter      0      1      2      3      4   ...   \n",
       "15     Letter      0      1      2      3      4   ...   \n",
       "16     Letter      0      1      2      3      4   ...   \n",
       "17     Letter      0      1      2      3      4   ...   \n",
       "18     Letter      0       1       2       3      4...   \n",
       "19     Letter      0      1      2      3      4   ...   \n",
       "20     Letter      0      1      2       3      4  ...   \n",
       "21     Letter      0       1       2       3       ...   \n",
       "22     Letter      0       1       2       3       ...   \n",
       "23     Letter       0       1       2       3      ...   \n",
       "24     Letter      0      1      2      3      4   ...   \n",
       "25     Letter      0       1       2       3       ...   \n",
       "26     Letter      0      1      2      3      4   ...   \n",
       "27     Letter      0      1      2      3      4   ...   \n",
       "28     Letter      0      1      2      3      4\n",
       "0 ...   \n",
       "29     Letter      0       1      2      3      4  ...   \n",
       "..                                                 ...   \n",
       "94     Letter      0      1      2      3      4   ...   \n",
       "95     Letter      0      1      2      3      4   ...   \n",
       "96     Letter       0       1       2      3       ...   \n",
       "97     Letter      0      1      2      3      4   ...   \n",
       "98     Letter      0      1       2      3       4 ...   \n",
       "99     Letter      0      1      2      3      4   ...   \n",
       "100    Letter      0      1      2      3      4   ...   \n",
       "101    Letter      0      1      2      3      4\n",
       "0 ...   \n",
       "102    Letter      0      1      2      3      4   ...   \n",
       "103    Letter      0      1      2      3      4   ...   \n",
       "104    Letter      0      1      2      3      4\n",
       "0 ...   \n",
       "105    Letter    0      1      2      3      4\n",
       "0   ...   \n",
       "106    Letter      0      1      2      3      4\n",
       "0 ...   \n",
       "107    Letter      0      1      2       3      4  ...   \n",
       "108    Letter      0      1      2    3    4    5  ...   \n",
       "109    Letter      0      1       2       3      4 ...   \n",
       "110    Letter       0       1       2       3      ...   \n",
       "111    Letter      0      1      2      3       4  ...   \n",
       "112    Letter      0       1       2      3       4...   \n",
       "113    Letter      0      1       2       3       4...   \n",
       "114    Letter       0       1       2       3      ...   \n",
       "115    Letter      0      1      2      3\n",
       "0      A ...   \n",
       "116    Letter      0       1       2       3       ...   \n",
       "117    Letter      0      1      2      3      4   ...   \n",
       "118    Letter      0      1      2      3      4   ...   \n",
       "119    Letter      0      1      2      3\n",
       "0      A ...   \n",
       "120    Letter      0      1      2      3      4   ...   \n",
       "121    Letter      0      1      2      3       4  ...   \n",
       "122    Letter      0      1      2      3      4\n",
       "0 ...   \n",
       "123    Letter      0      1      2      3      4   ...   \n",
       "\n",
       "                                           ReversePSSM       TF  \n",
       "0      Letter      0       1      2      3      4  ...     HSF1  \n",
       "1      Letter      0      1      2       3      4  ...     RPN4  \n",
       "2      Letter      0       1       2       3       ...     THI2  \n",
       "3      Letter      0      1    2    3    4    5    ...     UGA3  \n",
       "4      Letter      0      1      2      3      4   ...     SMP1  \n",
       "5      Letter       0      1       2       3       ...     RTG3  \n",
       "6      Letter      0      1      2      3       4  ...     CIN5  \n",
       "7      Letter      0      1      2      3      4   ...    DAL82  \n",
       "8      Letter      0      1      2      3\n",
       "0      A ...    DAL80  \n",
       "9      Letter      0      1      2      3      4   ...    DAL81  \n",
       "10     Letter      0      1      2      3      4   ...     REB1  \n",
       "11     Letter      0      1      2      3      4   ...     PUT3  \n",
       "12     Letter      0      1      2       3       4 ...     BAS1  \n",
       "13     Letter      0      1      2      3      4   ...     YAP3  \n",
       "14     Letter      0      1      2      3      4   ...     SIP4  \n",
       "15     Letter      0      1      2      3      4   ...     SOK2  \n",
       "16     Letter      0      1      2      3      4   ...     NDD1  \n",
       "17     Letter      0       1      2       3      4 ...     FKH1  \n",
       "18     Letter      0       1      2       3       4...     SKO1  \n",
       "19     Letter      0      1       2       3      4 ...     FKH2  \n",
       "20     Letter      0      1      2       3      4  ...     CBF1  \n",
       "21     Letter       0       1       2       3      ...     GCR1  \n",
       "22     Letter       0       1       2       3      ...     GCR2  \n",
       "23     Letter       0       1       2       3      ...     YOX1  \n",
       "24     Letter      0      1      2      3      4   ...     HAP1  \n",
       "25     Letter       0       1       2       3      ...     OPI1  \n",
       "26     Letter      0      1      2      3      4   ...     CRZ1  \n",
       "27     Letter      0      1      2      3      4   ...     ABF1  \n",
       "28     Letter      0      1      2      3      4\n",
       "0 ...  YDR026C  \n",
       "29     Letter      0      1      2      3       4  ...     SUT1  \n",
       "..                                                 ...      ...  \n",
       "94     Letter      0      1      2      3      4   ...     RPH1  \n",
       "95     Letter      0      1      2      3      4   ...     ASH1  \n",
       "96     Letter       0       1       2      3       ...     XBP1  \n",
       "97     Letter      0      1      2      3      4   ...     IME1  \n",
       "98     Letter      0      1      2       3       4 ...     SNT2  \n",
       "99     Letter      0      1      2      3      4   ...     MAC1  \n",
       "100    Letter      0      1      2      3      4   ...    MATA1  \n",
       "101    Letter      0      1      2      3      4\n",
       "0 ...     GZF3  \n",
       "102    Letter      0      1      2      3      4   ...     NRG1  \n",
       "103    Letter      0      1      2      3      4   ...  YML081W  \n",
       "104    Letter      0      1      2      3      4\n",
       "0 ...     YHP1  \n",
       "105    Letter      0      1      2      3    4\n",
       "0   ...     ADR1  \n",
       "106    Letter      0      1      2      3      4\n",
       "0 ...     GTS1  \n",
       "107    Letter      0      1      2       3      4  ...     UME6  \n",
       "108    Letter      0      1    2    3    4    5    ...    GAL80  \n",
       "109    Letter      0       1      2       3      4 ...     AZF1  \n",
       "110    Letter      0       1       2       3       ...     LEU3  \n",
       "111    Letter      0      1       2       3      4 ...     MCM1  \n",
       "112    Letter       0       1       2       3      ...    MET32  \n",
       "113    Letter       0       1       2       3      ...    MET31  \n",
       "114    Letter       0       1       2       3      ...     SNF1  \n",
       "115    Letter      0      1      2      3\n",
       "0      A ...     HAP5  \n",
       "116    Letter      0      1      2      3      4   ...     HAP4  \n",
       "117    Letter      0      1      2      3      4   ...     IXR1  \n",
       "118    Letter      0      1      2      3      4   ...     SFL1  \n",
       "119    Letter      0      1      2      3\n",
       "0      A ...     HAP3  \n",
       "120    Letter      0      1      2      3      4   ...     HAP2  \n",
       "121    Letter       0       1      2      3       4...     RAP1  \n",
       "122    Letter      0      1      2      3      4\n",
       "0 ...     MOT3  \n",
       "123    Letter      0      1      2      3      4   ...     SKN7  \n",
       "\n",
       "[124 rows x 3 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get the data from the tamo file that I'm working with\n",
    "tamoDF = getPSSMs('Data/yeast.tamo')\n",
    "# Get the amount of PSSMs (or TFs, both are the same)\n",
    "countTF = tamoDF.count()\n",
    "numOfTF = countTF['TF']\n",
    "# Show the dataframe\n",
    "tamoDF"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "We have just read in all of the PSSMs for each specific transcription factor. There is one more component that we need to implement before we're all set to go ahead and calculate all of the hits. That component is a function that takes in a sequence (that is the same length (horizontal length) as the PSSM) and a PSSM and calculates the output of the PSSM when you plug in that sequence. Let's go ahead and do that."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def outputOfPSSM(PSSM, sequence):\n",
    "    # Find the length (horizontal length) of the PSSM\n",
    "    lenOfPSSM = PSSM.count(1) - 1\n",
    "    lenOfPSSM = lenOfPSSM[0]\n",
    "    # Find the maximum possible output of the PSSM\n",
    "    maximumList = list(PSSM.max())\n",
    "    del maximumList[0]\n",
    "    maximum = 0\n",
    "    for i in range(len(maximumList)):\n",
    "        maximum += maximumList[i]\n",
    "    # Calculate the output of the PSSM\n",
    "    output = 0\n",
    "    for i in range(len(sequence)):\n",
    "        if sequence[i] == 'A':\n",
    "            output += PSSM[str(i)][0]\n",
    "        if sequence[i] == 'C':\n",
    "            output += PSSM[str(i)][1]\n",
    "        if sequence[i] == 'T':\n",
    "            output += PSSM[str(i)][2]\n",
    "        if sequence[i] == 'G':\n",
    "            output += PSSM[str(i)][3]\n",
    "    output /= maximum\n",
    "    # Return the output of the PSSM\n",
    "    return output"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now that we have all of the components, let's calculate the hits."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "def calculateHits(fastaDF, tamoDF, numOfTF, numOfCHR, weakThreshold, strongThreshold, Path):\n",
    "    # Get the time\n",
    "    t_beg = time()\n",
    "    # Check to see if the path is made (if not, make it)\n",
    "    if not os.path.exists(Path):\n",
    "        os.makedirs(Path)\n",
    "    # Calculate hits\n",
    "    for i in range(numOfTF):\n",
    "        pssm = tamoDF['ForwardsPSSM'][i]\n",
    "        REVpssm = tamoDF['ReversePSSM'][i]\n",
    "        TF = tamoDF['TF'][i]\n",
    "        lenOfPSSM = pssm.count(1) - 1\n",
    "        lenOfPSSM = lenOfPSSM[0]\n",
    "        filename = Path + '/Hits' + TF + '.gff'\n",
    "        fileID = open(filename, 'w')\n",
    "        for j in range(numOfCHR):\n",
    "            CHR = fastaDF['sequence'][j]\n",
    "            print('Calculating the hits for the transcription factor', TF, 'using chromosome', fastaDF['chromosome'][j])\n",
    "            t_intermediate_beg = time()\n",
    "            for k in range(len(CHR)):\n",
    "                if (k + lenOfPSSM) > len(CHR):\n",
    "                    output1 = outputOfPSSM(pssm, CHR[k:] + str([' ' for l in range(k + lenOfPSSM - 1 - len(CHR))]))\n",
    "                    output2 = outputOfPSSM(REVpssm, CHR[k:] + str([' ' for l in range(k + lenOfPSSM - 1 - len(CHR))]))\n",
    "                    end = len(CHR)\n",
    "                else:\n",
    "                    output1 = outputOfPSSM(pssm, CHR[k:(k + lenOfPSSM)])\n",
    "                    output2 = outputOfPSSM(REVpssm, CHR[k:(k + lenOfPSSM)])\n",
    "                    end = k + lenOfPSSM\n",
    "                if output1 >= weakThreshold:\n",
    "                    if output1 < strongThreshold:\n",
    "                        fileID.write(fastaDF['chromosome'][j] + '\\t' + TF + '\\t' + 'hit' + '\\t' + str(k) + '\\t' + str(end) + '\\t' + str(output1) + '\\t' + '+' + '\\t' + '.' + '\\t' + 'weak' + '\\n')\n",
    "                    else:\n",
    "                        fileID.write(fastaDF['chromosome'][j] + '\\t' + TF + '\\t' + 'hit' + '\\t' + str(k) + '\\t' + str(end) + '\\t' + str(output1) + '\\t' + '+' + '\\t' + '.' + '\\t' + 'strong' + '\\n')\n",
    "                if output2 >= weakThreshold:\n",
    "                    if output2 < strongThreshold:\n",
    "                        fileID.write(fastaDF['chromosome'][j] + '\\t' + TF + '\\t' + 'hit' + '\\t' + str(k) + '\\t' + str(end) + '\\t' + str(output2) + '\\t' + '-' + '\\t' + '.' + '\\t' + 'weak' + '\\n')\n",
    "                    else:\n",
    "                        fileID.write(fastaDF['chromosome'][j] + '\\t' + TF + '\\t' + 'hit' + '\\t' + str(k) + '\\t' + str(end) + '\\t' + str(output2) + '\\t' + '-' + '\\t' + '.' + '\\t' + 'strong' + '\\n')\n",
    "            t_intermediate_end = time()\n",
    "            print('It took', (t_intermediate_end - t_intermediate_beg)/60, 'minutes to calculate the hits for transcription factor', TF, 'using chromosome', fastaDF['chromosome'][j])\n",
    "        fileID.close()\n",
    "    # Get the time\n",
    "    t_end = time()\n",
    "    # Return the time it took to calculate the hits\n",
    "    return t_end - t_beg"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's talk about the reasoning behind this. I'm given the two dataframes that contain the fasta and the tamo data. I want to go through each trancription factor and calculate the hits for every single chromosome."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Calculating the hits for the transcription factor HSF1 using chromosome chr1\n",
      "It took 18.438492600123087 minutes to calculate the hits for transcription factor HSF1 using chromosome chr1\n",
      "Calculating the hits for the transcription factor HSF1 using chromosome chr2\n",
      "It took 117.31964531342189 minutes to calculate the hits for transcription factor HSF1 using chromosome chr2\n",
      "Calculating the hits for the transcription factor HSF1 using chromosome chr3\n",
      "It took 30.447948948542276 minutes to calculate the hits for transcription factor HSF1 using chromosome chr3\n",
      "Calculating the hits for the transcription factor HSF1 using chromosome chr4\n",
      "It took 168.77411204973856 minutes to calculate the hits for transcription factor HSF1 using chromosome chr4\n",
      "Calculating the hits for the transcription factor HSF1 using chromosome chr5\n",
      "It took 50.9314360499382 minutes to calculate the hits for transcription factor HSF1 using chromosome chr5\n",
      "Calculating the hits for the transcription factor HSF1 using chromosome chr6\n",
      "It took 25.224676716327668 minutes to calculate the hits for transcription factor HSF1 using chromosome chr6\n",
      "Calculating the hits for the transcription factor HSF1 using chromosome chr7\n",
      "It took 926.345464237531 minutes to calculate the hits for transcription factor HSF1 using chromosome chr7\n"
     ]
    }
   ],
   "source": [
    "time_est = calculateHits(fastaDF, tamoDF, numOfTF, numOfCHR, 0.35, 0.7, 'Hits Files')\n",
    "time_est"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Part 2: Analysing the Hits\n",
    "---\n",
    "We have already implemented a function to calculate the hits, given a transcription factor and a chromosome (or all of the chromosomes), we have to analyse these hits. I analyse the hits by "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "def hist(InFilename, HitsInFilename, PicFile):\n",
    "    # Get the time at the begining of the function\n",
    "    t_beg = time()\n",
    "    # Open the input file that holds the hits\n",
    "    fileID = open(HitsInFilename, 'r')\n",
    "    # Initialize the list that will hold the input\n",
    "    data = []\n",
    "    # Read in the input file\n",
    "    line = fileID.readline()\n",
    "    while line:\n",
    "        data.append(line[:len(line)-1])\n",
    "        line = fileID.readline()\n",
    "    # Parse the input\n",
    "    weakSites = []\n",
    "    strongSites = []\n",
    "    for i in range(len(data)):\n",
    "        line = data[i]\n",
    "        if len(line) != 0:\n",
    "            if line[0] != '#':\n",
    "                # Get the position of the hit\n",
    "                position = line[2:8]\n",
    "                position = int(position)\n",
    "                # Get the type of the hit\n",
    "                typeOfHit = line[len(line)-8:len(line)-2]\n",
    "                # Input the position into the correct list\n",
    "                if typeOfHit == 'strong':\n",
    "                    strongSites.append(position)\n",
    "                if typeOfHit == '  weak':\n",
    "                    weakSites.append(position)\n",
    "    # Take out all of the duplicates in the lists holding the hits\n",
    "#     newList = []\n",
    "#     for i in range(len(weakSites)):\n",
    "#         target = weakSites[i]\n",
    "#         found = False\n",
    "#         i = 0\n",
    "#         while i < len(newList) and found == False:\n",
    "#             if target == newList[i]:\n",
    "#                 found = True\n",
    "#             else:\n",
    "#                 i += 1\n",
    "#         if found == False:\n",
    "#             newList.append(target)\n",
    "#     weakSites = newList\n",
    "#     newList = []\n",
    "#     for i in range(len(strongSites)):\n",
    "#         target = strongSites[i]\n",
    "#         found = False\n",
    "#         i = 0\n",
    "#         while i < len(newList) and found == False:\n",
    "#             if target == newList[i]:\n",
    "#                 found = True\n",
    "#             else:\n",
    "#                 i += 1\n",
    "#         if found == False:\n",
    "#             newList.append(target)\n",
    "#     strongSites = newList\n",
    "    #weakSites = list(set(weakSites))\n",
    "    #weakSites.sort()\n",
    "    #strongSites = list(set(strongSites))\n",
    "    #strongSites.sort()\n",
    "    # Close the input file that holds the hits\n",
    "    fileID.close()\n",
    "    # Read in the input file\n",
    "    Input = readInInput(InFilename)\n",
    "    # Get the number of plus or minus range of what we're going to be anaylsing and the size of the buckets\n",
    "    r = Input[5]\n",
    "    s = Input[4]\n",
    "    # Analyse the hits\n",
    "    categories = [0 for i in range(r//s)]\n",
    "#     for i in range(len(strongSites)):\n",
    "#         for j in range(len(weakSites)):\n",
    "#             value = abs(strongSites[i] - weakSites[j])\n",
    "#             for k in range(r//s):\n",
    "#                 if (value <= (s*k) and (value > s*(k-1))):\n",
    "#                     categories[k] += 1\n",
    "    for i in range(len(strongSites)):\n",
    "        position1 = 0\n",
    "        position2 = 0\n",
    "        for j in range(len(weakSites)):\n",
    "            if weakSites[j] <= strongSites[i]-r:\n",
    "                position1 = j\n",
    "            if weakSites[j] <= strongSites[i]+r:\n",
    "                position2 = j\n",
    "        for j in range(position1, position2):\n",
    "            value = abs(strongSites[i] - weakSites[j])\n",
    "            for k in range(r//s):\n",
    "                if (value <= (s*k)) and (value > s*(k-1)):\n",
    "                    categories[k] += 1\n",
    "    # Plot the histogram\n",
    "    fig, ax = plt.subplots(nrows=1, ncols=1, figsize=(10,5))\n",
    "    ax.bar(range(len(categories)), categories, color='black', label='Histogram')\n",
    "    plt.plot(range(len(categories)), [mean(categories) for i in range(len(categories))], color='red', label='Mean')\n",
    "    plt.plot(range(len(categories)), [2*std(categories)+mean(categories) for i in range(len(categories))], color='steelblue', label='Standard Deviation (+2)')\n",
    "    plt.plot(range(len(categories)), [std(categories)+mean(categories) for i in range(len(categories))], color='cyan', label='Standard Deviation (+1)')\n",
    "    if mean(categories)-std(categories) >= 0:\n",
    "        plt.plot(range(len(categories)), [mean(categories)-std(categories) for i in range(len(categories))], color='cyan', label='Standard Deviation (-1)')\n",
    "    if mean(categories)-2*std(categories) >= 0:\n",
    "        plt.plot(range(len(categories)), [mean(categories)-2*std(categories) for i in range(len(categories))], color='steelblue', label='Standard Deviation (-2)')\n",
    "    ax.legend()\n",
    "    path = ''\n",
    "    for i in range(len(PicFile)):\n",
    "        if PicFile[i] != '/':\n",
    "            path += PicFile[i]\n",
    "        else:\n",
    "            break\n",
    "    if not os.path.exists(path):\n",
    "        os.makedirs(path)\n",
    "    fig.savefig(PicFile)\n",
    "    plt.show()\n",
    "    # Get the time at the end of the function\n",
    "    t_end = time()\n",
    "    # Return the amount of seconds it took the function to run (in seconds)\n",
    "    return t_end - t_beg"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "Now that we have "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "# Import the function that gets the time\n",
    "from time import time\n",
    "\n",
    "def main(InFilename):\n",
    "    # Get the time at the begining of the function\n",
    "    t_beg = time()\n",
    "    \n",
    "    Input = readInInput(InFilename)\n",
    "    fileID = open(Input[1])\n",
    "    line = fileID.readline()\n",
    "    while line:\n",
    "        if line[:6] == 'Source':\n",
    "            print('Transcription Factor:', line[9:len(line)-1])\n",
    "            time_est = calculateHits(Input[0], 'all', Input[1], line[9:len(line)-1], 'Hits/hits' + line[9:len(line)-1] + '.txt')\n",
    "            print('It took', time_est, 'seconds to calculate the hits')\n",
    "            time_est = hist(InFilename, 'Hits/hits' + line[9:len(line)-1] + '.txt', 'Histograms/histogram' + line[9:len(line)-1] + '.png')\n",
    "            print('It took', time_est, 'seconds to analyse the hits')\n",
    "        line = fileID.readline()\n",
    "    fileID.close()  \n",
    "    # Get the time at the end of the function\n",
    "    t_end = time()\n",
    "    # Return the amount of seconds it took the function to run (in seconds)\n",
    "    return t_end - t_beg\n",
    "\n",
    "main('Data/input.txt')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
